### PR TITLE
#1860: Validate on blur and memoize some things

### LINKED ISSUE #1860: Schema field changes slow down V2 brick keypress handling in dev build
Steps to reproduce
- Select v2 extension
- Add the new "Tour" brick
- Click add item
- Start typing into the "intro" field

### PR DESCRIPTION
**Pull request by BLoe**
Fixes #1860 

The biggest change here is to changer our Formik setup so that we're validating inputs on blur events instead of on change events. This was, by a long shot, the most effective change I made that helped typing speed in inputs in dev builds of the page editor. I also memo-ized a few components and tried to make sure that their props were being memo-ized properly.

### CODE REVIEW

File: src/components/fields/schemaFields/widgets/ArrayWidget.tsx [hunk 1]
```diff
@@ -71,7 +72,11 @@ const ArrayWidget: React.FC<SchemaFieldProps> = ({ schema, name }) => {
     throw new TypeError("Schema required for items");
   }
 
-  const schemaItems = schema.items ?? { additionalProperties: true };
+  const schemaItems = useMemo<JSONSchema7>(
```
Comment (twschiller) [inline src/components/fields/schemaFields/widgets/ArrayWidget.tsx:75]: We've been using Schema from core.ts to punt the JSONSchema versioning question down the road

File: src/devTools/editor/panes/EditorPane.tsx [hunk 2]
```diff
@@ -67,6 +67,8 @@ const EditorPane: React.FunctionComponent<{
               "Formik's submit should not be called to save an extension. Use 'saveElement' from 'useSavingWizard' instead."
             );
           }}
+          validateOnChange={false}
```
Comment (twschiller) [inline src/devTools/editor/panes/EditorPane.tsx:70]: Let's add a comment here with context for why we're setting it up this way

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx [hunk 3]
```diff
@@ -58,6 +58,56 @@ const BlockConfiguration: React.FunctionComponent<{
 
   const advancedOptionsRef = useRef<HTMLDivElement>();
 
+  const templateEngineOptions = useMemo<Array<Option<TemplateEngine>>>(
```
Comment (twschiller) [inline src/devTools/editor/tabs/effect/BlockConfiguration.tsx:61]: These don't change, so let's move them to the module level

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx [hunk 4]
```diff
@@ -58,6 +58,56 @@ const BlockConfiguration: React.FunctionComponent<{
...
+    () => [
+      { label: "Mustache", value: "mustache" },
+      { label: "Handlebars", value: "handlebars" },
+      { label: "Nunjucks", value: "nunjucks" },
+    ],
+    []
+  );
+
+  const templateEngineDescription = useMemo(
```
Comment (twschiller) [inline src/devTools/editor/tabs/effect/BlockConfiguration.tsx:70]: This doesn;t change, so let's move them to the module level

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx [hunk 5]
```diff
@@ -58,6 +58,56 @@ const BlockConfiguration: React.FunctionComponent<{
...
+        <code>true</code>, <code>t</code>, <code>yes</code>, <code>y</code>,{" "}
+        <code>on</code>, and <code>1</code> (case-insensitive)
+      </p>
+    ),
+    []
+  );
+
+  const targetOptions = useMemo<Array<Option<BlockWindow>>>(
+    () => [
+      { label: "Current Tab (self)", value: "self" },
```
Comment (twschiller) [inline src/devTools/editor/tabs/effect/BlockConfiguration.tsx:102]: This doesn;t change, so let's move them to the module level

**Comment by twschiller** [comment]
I'm good with this once @BALEHOK approves

**Comment by twschiller** [comment]
Adding @mnholtz as FYI on memoization

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx (hunk 3 above)
Comment (BLoe) [reply src/devTools/editor/tabs/effect/BlockConfiguration.tsx:61]: I caught this in the next PR I'm working on for the toggles on custom block configs that we discussed today (See #1860 ). I already made the changes on that branch.

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx (hunk 4 above)
Comment (BLoe) [reply src/devTools/editor/tabs/effect/BlockConfiguration.tsx:70]: (Same as the other one, I fixed this in my next PR)

File: src/devTools/editor/tabs/effect/BlockConfiguration.tsx (hunk 5 above)
Comment (BLoe) [reply src/devTools/editor/tabs/effect/BlockConfiguration.tsx:102]: (Same as the other one, I fixed this in my next PR)

**Review by BALEHOK — APPROVED** [review]

**Pull request merged**